subsys/shell/shell_uart.c

```diff
@@ -41,12 +42,33 @@ static void uart_rx_handle(const struct shell_uart *sh_uart)
 		if (len) {
 			rd_len = uart_fifo_read(sh_uart->ctrl_blk->dev,
 						data, len);
+#ifdef CONFIG_UART_CONSOLE_MCUMGR
+			/* Divert bytes from normal console handling if it is
+			 * part of an mcumgr frame.
+			 */
+			size_t i;
+
+			for (i = 0; i < rd_len; i++) {
+				if (!smp_shell_rx_byte(&ctx->smp, data[i])) {
+					break;
+				}
+			}
+			rd_len -= i;
+			if (rd_len) {
+				new_data = true;
+				for (u32_t j = 0; j < rd_len; j++) {
+					data[j] = data[i + j];
+				}
+			}
+			ring_buf_put_finish(sh_uart->rx_ringbuf, rd_len);
+#else
 
 			if (rd_len) {
 				new_data = true;
 			}
 
 			ring_buf_put_finish(sh_uart->rx_ringbuf, rd_len);
+#endif          /* CONFIG_UART_CONSOLE_MCUMGR */
```

Comment:
you can move `#endif` one line up and remove `ring_buf_put_finish` function call in line 63